### ISSUE TITLE
Restoring values from variables to sliders and switches (maybe something else) does not work

### ISSUE THREAD

**Issue by mrheadyhead** [thunkable/thunkable-issues#572]
### Platform
✕, ✕ Android companion

### Issue/Request Origin
Restoring values from variables to sliders and switches (maybe something else) does not work

### Steps to Reproduce
1. Save the value into a global variable, for example from a slider or switch
2. Chenge position of slider or switch
3. Restore value from global variable

### Project Link
https://x.thunkable.com/copy/8b3190895c6a1f755fe6170dd71290d0

### Expected Behavior
Position of the thumb of the sliders/switches changes, actual value of sliders equal to restored values from global varieable, functions are performed inside the sliders/switches blocks.

### Actual Behavior
Only the position of the thumb of the sliders/switches changes, but not the actual value of sliders. Accordingly, no functions are performed inside the sliders/switches blocks.

In Thunkable Classic everything worked as expected
```
  STACK TRACE AND/OR SCREENSHOTS
```

**Comment by MrMazzone** [comment]
Hi @mrheadyhead ,

There is nothing wrong with Thunkable here.  You simply forgot to update the label values when you "load" them.

Working version: https://x.thunkable.com/copy/17e0df48e4238a9ea213b40474b280f4

I suggest using the Thunkable Community site (https://community.thunkable.com/) as you learn more about Thunkable X and need help troubleshooting code.

**Comment by mrheadyhead** [comment]
@MrMazzone  Thank you for your feedback, but your option also does not work, moreover, it doubles the number of blocks, which in the future will lead to program lags when using a large number of controls.

I know about Community site - no one answered my question there.

Here is a screenshot frome classic version, and everything worked fine without re-reading the values from the sliders
![Annotation 2020-06-24 070642](https://user-images.githubusercontent.com/67040158/85499079-9c268900-b5e9-11ea-97ed-105d86edcf8c.jpg)

**Comment by MrMazzone** [comment]
@mrheadyhead I didn't see it do that previously.  Sadly, the solution to your problem does require an additional block. This is a more recent issue discussed on GitHub.  Adding a wait block (even just for 0 seconds) between the value change and the showing it on the label corrects this issue. However, nobody has reported it it increasing program lag, etc.
![image](https://user-images.githubusercontent.com/40954645/85577450-3c3edb00-b607-11ea-82ed-ffb60caf4a73.png)

I will preface transitioning from old Thunkable (MIT App Inventor based) to new Thunkable X this way -- it is different! They had to make it different to allow for cross-platform compatibility.  Which is a HUGE plus.  It may be frustrating at first, but once you understand how X works, I much prefer it to the old Thunkable/MIT App Inventor.

As for what you displayed above, here is my solution using the Local_Storage component in Thunkable X. Seems to work great.  Take a look.
https://x.thunkable.com/copy/845c0902554a97492bd6bf9cd1e0147e

**Comment by mrheadyhead** [comment]
@MrMazzone 
Thank you very much for your help. Now everything works.

I’ll write a little more about my issue report  - it seems logical to me, that after a variable has been read from the starage and assigned to the slider, it is necessary to perform all the functions that are contained in the block of this slider, just as if the slider had been shifted by a finger - that's why I opened the issue report. Now we have to create additional blocks that essentially repeat the functionality of existing ones.

Thunkable X is certainly better, and I hope that it will continue to develop, and the errors will be fixed

**Comment by MrMazzone** [comment]
@mrheadyhead You are very welcome.  

I am still a bit confused by what you want.  I think you want to be able to assign a variable to the slider value and have it dynamically update as the slider changes value.  This did not happen in Thunkable Classic though. However, there is a slider "value" property that is tied to its value.  You can get or set that property (as seen in my example).
![image](https://user-images.githubusercontent.com/40954645/85611645-255bb100-b626-11ea-9254-e6ce43e5b4cf.png)

If you are referring to the Label dynamically updating with the change of the slider, I also do not remember this from Thunkable Classic.  If you change the slider value, you must change the Label's text to reflect that change. 

Maybe if you can explain what you are looking for in more detail, I can understand this ticket.

**Comment by mrheadyhead** [comment]
@MrMazzone 
English is not my native language, so it may be difficult for you to understand me.
Once again, I want that when assigning a value to the slider, after pressing the button, blocks ((no matter which, in this case label) are marked as 2) are executed immediately inside the slider block (marked as 1).
![Annotation 2020-06-25 075933](https://user-images.githubusercontent.com/67040158/85657226-70270880-b6ba-11ea-880b-3bbe775f5c23.jpg)
Here is exactly the same example in the classic version, and it works great
![Annotation 2020-06-25 074724](https://user-images.githubusercontent.com/67040158/85657700-065b2e80-b6bb-11ea-856b-59c7fef8bbb2.jpg)
the problem is not the label, the fact is that the blocks are not executed automatically when assigning a value to the slider. For example, I need that after pressing "load", the block for sending data on the bluetooth, which is in the slider block, is executed, but this does not happen in version X. In the classic version, everything works.

Here is exactly the same program from the classic version, and it works

[Values_store.zip](https://github.com/thunkable/thunkable-issues/files/4829211/Values_store.zip)

**Comment by MrMazzone** [comment]
OK @mrheadyhead now I understand!

You want the Slider in Thunkable X's onValueChange event to trigger whenever the value changes (example, using property setter) not just when the user manually moves the thumb position.  This would be similar functionality to Thunkable Classic's Slider PositionChanged event.

I would recommend renaming this issue report "Slider OnValueChange event not triggered when value changes using property setter, only when user manually moves thumb position"

Thank you for being patient with me.  I hope I have been helpful, and hopefully the Thunkable team will review your issue and implement the change.

**Comment by josmas** [comment]
Hi all, can I ask, is this just for Android? has anyone tried web and iOS? Thanks!

**Comment by mrheadyhead** [comment]
@josmas Hi! Unfortunately, I do not have any iOs devices at the moment for the tests

**Comment by MrMazzone** [comment]
@josmas @mrheadyhead 
Just tested on iOS and web.  Same issue, the "on Value Change" event is not triggered when the value changes.  Only when the user manually moves thumb.

**Comment by josmas** [comment]
Thanks, all. I am not sure what the right behavior here is. In theory, setting a value is sort of a `reset operation`, it is like giving the component a value for the first time, and viewed like that, it should not call the change block.
My concern is that some people might not want onChange called the first time they set a value.

I guess if you have the blocks you call inside onChange in a function, you can call the function when you are setting the block as well as when it gets automatically called by onChange. What do you all think?

**Comment by MrMazzone** [comment]
I agree @josmas.  Having it not automatically change gives the Thunkable users more control over behavior.  Forcing it to trigger the event could break other people's apps and prevents someone from using the event in the way it works now.  A function can certainly help manage the complexity with apps that change the values often, like what @mrheadyhead has. 

My recommendation is maybe renaming the event handler?  Maybe the name doesn't properly represent the when the event is triggered?  A better name would be "on Thumb Change" or "on Slider Dragged".  Similar change for Switch would work too.

**Comment by mrheadyhead** [comment]
@josmas Hi! Maybe you can add another separate event that will execute the internal blocks.
And a separate event that will only move the thumb

**Comment by domhnallohanlon** [comment]
Hey @mrheadyhead, agreed it would be great to have this in the blocks too, but I was able to create the functionality that you were looking for with a variable and a function.

These blocks work when the switch (or slider) is changed by hand and when it is changed via the button click.

<img width="605" alt="Screenshot 2021-02-12 at 15 13 59" src="https://user-images.githubusercontent.com/3117345/107785798-f032d900-6d44-11eb-9c46-df78fa0b0a0a.png">

Let me know whether or not that works for you. 

cc: @MrMazzone
